ideascube/serveradmin/catalog.py

```diff
@@ -0,0 +1,576 @@
+
+    @property
+    def name(self):
+        return self.__class__.__name__.lower()
+
+    def install(self, package, download_path):
+        package.install(download_path, self._install_dir)
+        self._installed.append(package)
+
+    def remove(self, package):
+        package.remove(self._install_dir)
+        self._removed.append(package)
+
+    def commit(self):
+        self._installed = []
+        self._removed = []
+
+    def restart_service(self, name):
+        sys.stdout.write('Restarting service "{}"\n'.format(name))
+        try:
+            manager = SystemManager()
+            service = manager.get_service(name)
+
+        except NoSuchUnit:
+            # Service is not installed, give up.
+            pass
+
+        else:
+            manager.restart(service.Id)
+
+
+class Kiwix(Handler):
+
+    def commit(self):
+        sys.stdout.write('Rebuilding the Kiwix library\n')
+        library = etree.Element('library')
+        libdir = os.path.join(self._install_dir, 'data', 'library')
+
+        for libpath in glob(os.path.join(libdir, '*.xml')):
+            zimname = os.path.basename(libpath)[:-4]
+
+            with open(libpath, 'r') as f:
+                et = etree.parse(f)
+                books = et.findall('book')
+
+                # Messing with the path gets complicated otherwise
+                # TODO: Can we assume this is always true for stuff distributed
+                #       by kiwix?
+                assert len(books) == 1
+
+                book = books[0]
+                book.set('path', 'data/content/%s' % zimname)
+                book.set('indexPath', 'data/index/%s.idx' % zimname)
+
+                library.append(book)
+
+        with open(os.path.join(self._install_dir, 'library.xml'), 'wb') as f:
+            f.write(etree.tostring(
+                library, xml_declaration=True, encoding='utf-8'))
+
+        self.restart_service('kiwix-server')
+        super().commit()
+
+
+class Nginx(Handler):
+    template = """
+server {{
+    listen   80;
+    server_name {server_name};
+    root {root};
+    index index.html;
+}}
+"""
+    root = '/etc/nginx/'
+
+    def commit(self):
+        for pkg in self._removed:
+            self.unregister_site(pkg)
+        for pkg in self._installed:
+            self.register_site(pkg)
+        self.restart_service('nginx')
+        super().commit()
+
+    def register_site(self, package):
+        available = Path(self.root, 'sites-available', package.id)
+        enabled = Path(self.root, 'sites-enabled', package.id)
+        server_name = '{subdomain}.{domain}'.format(subdomain=package.id,
+                                                    domain=settings.DOMAIN)
+        root = package.get_root_dir(self._install_dir)
+        with available.open('w') as f:
+            f.write(self.template.format(server_name=server_name, root=root))
+        try:
+            enabled.symlink_to(available)
+        except FileExistsError:
+            pass
```

Comment:
Not sure though what to do when the target is not the expected one: in theory it can even be a folder with content we would not want to delete without the user agreement. Or it can be pointing to another site we want to keep (id clash between two catalog for example).
I've update the code to raise when the symlink exist and it's not pointing to the expected target, but that will need refinement for sure.